File: problems/problemset/tiny-interpreter/Python/solution/interpreter.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any

# ── Tokens ────────────────────────────────────────────────────────────────────

NUM, IDENT = 'NUM', 'IDENT'
KEYWORDS = {'if', 'elif', 'else', 'while', 'def', 'return', 'print', 'and', 'or', 'not'}
NEWLINE, INDENT, DEDENT, END = 'NEWLINE', 'INDENT', 'DEDENT', 'END'

def lex(src: str):
    tokens = []
    indents = [0]

    for line in src.splitlines():
        first = len(line) - len(line.lstrip(' \t'))
        stripped = line[first:]
        if not stripped or stripped.startswith('#'):
            continue

        ind = first
        if ind > indents[-1]:
            indents.append(ind)
            tokens.append((INDENT, ''))
        else:
            while ind < indents[-1]:
                indents.pop()
                tokens.append((DEDENT, ''))

        i = 0
        while i < len(stripped):
            c = stripped[i]
            if c in ' \t':
                i += 1; continue
            if c == '#':
                break
            if c.isdigit():
                j = i
                while i < len(stripped) and stripped[i].isdigit():
                    i += 1
                tokens.append((NUM, int(stripped[j:i]))); continue
            if c.isalpha() or c == '_':
                j = i
                while i < len(stripped) and (stripped[i].isalnum() or stripped[i] == '_'):
                    i += 1
                word = stripped[j:i]
                tokens.append((word if word in KEYWORDS else IDENT, word)); continue
            two = stripped[i:i+2]
            if two in ('==', '!=', '<=', '>='):
                tokens.append((two, two)); i += 2; continue
            tokens.append((c, c)); i += 1

        tokens.append((NEWLINE, ''))

    while len(indents) > 1:
        indents.pop()
        tokens.append((DEDENT, ''))
    tokens.append((END, ''))
    return tokens

# ── AST nodes (plain tuples for speed) ───────────────────────────────────────
# Expr: ('num', v) | ('var', name) | ('bin', op, l, r) | ('unary', op, e) | ('call', name, args)
# Stmt: ('assign', name, expr) | ('print', expr) | ('return', expr)
#     | ('if', [(cond, block)], else_block) | ('while', cond, block)
#     | ('def', name, params, block) | ('expr', expr)

# ── Parser ────────────────────────────────────────────────────────────────────

class Parser:
    def __init__(self, tokens):
        self.tok = tokens
        self.pos = 0

    def cur(self):  return self.tok[self.pos][0]
    def val(self):  return self.tok[self.pos][1]
    def peek(self): return self.tok[self.pos+1][0] if self.pos+1 < len(self.tok) else END
    def consume(self):
        t = self.tok[self.pos]; self.pos += 1; return t
    def expect(self, t):
        assert self.cur() == t, f"expected {t} got {self.cur()}"
        return self.consume()
    def skip_nl(self):
        while self.cur() == NEWLINE: self.pos += 1

    def parse(self):
        block = []
        self.skip_nl()
        while self.cur() != END:
            block.append(self.stmt())
            self.skip_nl()
        return block

    def block(self):
        self.expect(INDENT)
        stmts = []
        while self.cur() not in (DEDENT, END):
            stmts.append(self.stmt())
        if self.cur() == DEDENT: self.pos += 1
        return stmts

    def stmt(self):
        self.skip_nl()
        c = self.cur()
        if c == 'if':
            return self.if_stmt()
        if c == 'while':
            self.pos += 1
            cond = self.expr(); self.expect(':')
            if self.cur() == NEWLINE: self.pos += 1
            return ('while', cond, self.block())
        if c == 'def':
            self.pos += 1
            name = self.expect(IDENT)[1]
            self.expect('(')
            params = []
            while self.cur() != ')':
                params.append(self.expect(IDENT)[1])
                if self.cur() == ',': self.pos += 1
            self.expect(')'); self.expect(':')
            if self.cur() == NEWLINE: self.pos += 1
            return ('def', name, params, self.block())
        if c == 'return':
            self.pos += 1
            e = self.expr()
            if self.cur() == NEWLINE: self.pos += 1
            return ('return', e)
        if c == 'print':
            self.pos += 1; self.expect('(')
            e = self.expr(); self.expect(')')
            if self.cur() == NEWLINE: self.pos += 1
            return ('print', e)
        if c == IDENT and self.peek() == '=':
            name = self.val(); self.pos += 2
            e = self.expr()
            if self.cur() == NEWLINE: self.pos += 1
            return ('assign', name, e)
        e = self.expr()
        if self.cur() == NEWLINE: self.pos += 1
        return ('expr', e)

    def if_stmt(self):
        self.pos += 1
        branches = []
        cond = self.expr(); self.expect(':')
        if self.cur() == NEWLINE: self.pos += 1
        branches.append((cond, self.block()))
        while self.cur() == 'elif':
            self.pos += 1
            c = self.expr(); self.expect(':')
            if self.cur() == NEWLINE: self.pos += 1
            branches.append((c, self.block()))
        else_block = []
        if self.cur() == 'else':
            self.pos += 1; self.expect(':')
            if self.cur() == NEWLINE: self.pos += 1
            else_block = self.block()
        return ('if', branches, else_block)

    def expr(self):   return self.or_expr()
    def or_expr(self):
        l = self.and_expr()
        while self.cur() == 'or':
            self.pos += 1; r = self.and_expr()
            l = ('bin', 'or', l, r)
        return l
    def and_expr(self):
        l = self.not_expr()
        while self.cur() == 'and':
            self.pos += 1; r = self.not_expr()
            l = ('bin', 'and', l, r)
        return l
    def not_expr(self):
        if self.cur() == 'not':
            self.pos += 1; return ('unary', 'not', self.not_expr())
        return self.cmp()
    def cmp(self):
        l = self.add()
        if self.cur() in ('==','!=','<','<=','>','>='):
            op = self.consume()[0]; return ('bin', op, l, self.add())
        return l
    def add(self):
        l = self.mul()
        while self.cur() in ('+', '-'):
            op = self.consume()[0]; l = ('bin', op, l, self.mul())
        return l
    def mul(self):
        l = self.unary()
        while self.cur() in ('*', '/', '%'):
            op = self.consume()[0]; l = ('bin', op, l, self.unary())
        return l
    def unary(self):
        if self.cur() == '-':
            self.pos += 1; return ('unary', '-', self.unary())
        return self.primary()
    def primary(self):
        c = self.cur()
        if c == NUM:
            return ('num', self.consume()[1])
        if c == IDENT:
            name = self.consume()[1]
            if self.cur() == '(':
                self.pos += 1
                args = []
                while self.cur() != ')':
                    args.append(self.expr())
                    if self.cur() == ',': self.pos += 1
                self.pos += 1
                return ('call', name, args)
            return ('var', name)
        if c == '(':
            self.pos += 1; e = self.expr(); self.expect(')'); return e
        return ('num', 0)

# ── Interpreter ───────────────────────────────────────────────────────────────

class ReturnSignal(Exception):
    def __init__(self, val): self.val = val

class Interpreter:
    def run(self, source: str) -> str:
        self._output = []
        self._funcs = {}
        tokens = lex(source)
        program = Parser(tokens).parse()
        for s in program:
            if s[0] == 'def':
                self._funcs[s[1]] = s
        try:
            self._exec_block(program, {})
        except ReturnSignal:
            pass
        return ''.join(self._output)

    def _exec_block(self, block, env):
        for s in block:
            self._exec(s, env)

    def _exec(self, s, env):
        kind = s[0]
        if kind == 'assign':
            env[s[1]] = self._eval(s[2], env)
        elif kind == 'print':
            self._output.append(str(self._eval(s[1], env)) + '\n')
        elif kind == 'return':
            raise ReturnSignal(self._eval(s[1], env))
        elif kind == 'if':
            for cond, body in s[1]:
                if self._eval(cond, env):
                    self._exec_block(body, env); return
            self._exec_block(s[2], env)
        elif kind == 'while':
            while self._eval(s[1], env):
                self._exec_block(s[2], env)
        elif kind == 'def':
            self._funcs[s[1]] = s
        elif kind == 'expr':
            self._eval(s[1], env)

    def _eval(self, e, env):
        kind = e[0]
        if kind == 'num':   return e[1]
        if kind == 'var':   return env.get(e[1], self._funcs.get(e[1], 0)) if e[1] not in self._funcs else env.get(e[1], 0)
        if kind == 'unary':
            v = self._eval(e[2], env)
            return -v if e[1] == '-' else (0 if v else 1)
        if kind == 'bin':
            op = e[1]
            if op == 'and':
                l = self._eval(e[2], env); return self._eval(e[3], env) if l else 0
            if op == 'or':
                l = self._eval(e[2], env); return l if l else self._eval(e[3], env)
            l, r = self._eval(e[2], env), self._eval(e[3], env)
            if op == '+':  return l + r
            if op == '-':  return l - r
            if op == '*':  return l * r
            if op == '/':  return int(l / r)
            if op == '%':  return l % r
            if op == '==': return 1 if l == r else 0
            if op == '!=': return 1 if l != r else 0
            if op == '<':  return 1 if l < r else 0
            if op == '<=': return 1 if l <= r else 0
            if op == '>':  return 1 if l > r else 0
            if op == '>=': return 1 if l >= r else 0
        if kind == 'call':
            fn = self._funcs.get(e[1])
            if fn is None: return 0
            # fn = ('def', name, params, body)  e = ('call', name, args)
            local = {p: self._eval(a, env) for p, a in zip(fn[2], e[2])}
            try:
                self._exec_block(fn[3], local)
            except ReturnSignal as ret:
                return ret.val
            return 0
        return 0

File: problems/problemset/tiny-interpreter/Rust/solution/interpreter.rs
use std::collections::HashMap;

// ── Tokens ────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq)]
enum TT {
    Num(i64), Ident(String),
    If, Elif, Else, While, Def, Return, Print, And, Or, Not,
    Plus, Minus, Star, Slash, Percent,
    Eq, Neq, Lt, Le, Gt, Ge, Assign,
    LParen, RParen, Comma, Colon,
    Newline, Indent, Dedent, End,
}

fn lex(src: &str) -> Vec<TT> {
    let mut out = Vec::new();
    let mut indents: Vec<usize> = vec![0];

    for line in src.lines() {
        let first = line.len() - line.trim_start_matches(|c| c == ' ' || c == '\t').len();
        let stripped = &line[first..];
        if stripped.is_empty() || stripped.starts_with('#') { continue; }

        let ind = first;
        if ind > *indents.last().unwrap() {
            indents.push(ind);
            out.push(TT::Indent);
        } else {
            while ind < *indents.last().unwrap() {
                indents.pop();
                out.push(TT::Dedent);
            }
        }

        let bytes = stripped.as_bytes();
        let mut i = 0;
        while i < bytes.len() {
            let c = bytes[i];
            if c == b' ' || c == b'\t' { i += 1; continue; }
            if c == b'#' { break; }

            if c.is_ascii_digit() {
                let start = i;
                while i < bytes.len() && bytes[i].is_ascii_digit() { i += 1; }
                let n: i64 = stripped[start..i].parse().unwrap();
                out.push(TT::Num(n));
                continue;
            }
            if c.is_ascii_alphabetic() || c == b'_' {
                let start = i;
                while i < bytes.len() && (bytes[i].is_ascii_alphanumeric() || bytes[i] == b'_') { i += 1; }
                let word = &stripped[start..i];
                out.push(match word {
                    "if"     => TT::If,   "elif"   => TT::Elif,   "else"   => TT::Else,
                    "while"  => TT::While, "def"   => TT::Def,    "return" => TT::Return,
                    "print"  => TT::Print, "and"   => TT::And,    "or"     => TT::Or,
                    "not"    => TT::Not,
                    _        => TT::Ident(word.to_string()),
                });
                continue;
            }
            if i + 1 < bytes.len() {
                let two = &stripped[i..i+2];
                match two {
                    "==" => { out.push(TT::Eq);  i += 2; continue; }
                    "!=" => { out.push(TT::Neq); i += 2; continue; }
                    "<=" => { out.push(TT::Le);  i += 2; continue; }
                    ">=" => { out.push(TT::Ge);  i += 2; continue; }
                    _ => {}
                }
            }
            out.push(match c {
                b'+' => TT::Plus,   b'-' => TT::Minus,  b'*' => TT::Star,
                b'/' => TT::Slash,  b'%' => TT::Percent, b'=' => TT::Assign,
                b'<' => TT::Lt,     b'>' => TT::Gt,
                b'(' => TT::LParen, b')' => TT::RParen,
                b',' => TT::Comma,  b':' => TT::Colon,
                _    => { i += 1; continue; }
            });
            i += 1;
        }
        out.push(TT::Newline);
    }
    while indents.len() > 1 { indents.pop(); out.push(TT::Dedent); }
    out.push(TT::End);
    out
}

// ── AST ───────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
enum Expr {
    Num(i64),
    Var(String),
    Bin(String, Box<Expr>, Box<Expr>),
    Unary(String, Box<Expr>),
    Call(String, Vec<Expr>),
}

#[derive(Debug, Clone)]
enum Stmt {
    Assign(String, Expr),
    Print(Expr),
    Return(Expr),
    If(Vec<(Expr, Vec<Stmt>)>, Vec<Stmt>),
    While(Expr, Vec<Stmt>),
    Def(String, Vec<String>, Vec<Stmt>),
    Expr(Expr),
}

// ── Parser ────────────────────────────────────────────────────────────────────

struct Parser { tok: Vec<TT>, pos: usize }

impl Parser {
    fn cur(&self) -> &TT  { &self.tok[self.pos] }
    fn peek(&self) -> &TT { self.tok.get(self.pos + 1).unwrap_or(&TT::End) }
    fn consume(&mut self) -> TT { let t = self.tok[self.pos].clone(); self.pos += 1; t }
    fn expect(&mut self, t: &TT) { assert!(std::mem::discriminant(self.cur()) == std::mem::discriminant(t)); self.pos += 1; }
    fn skip_nl(&mut self) { while *self.cur() == TT::Newline { self.pos += 1; } }

    fn parse(&mut self) -> Vec<Stmt> {
        let mut out = vec![];
        self.skip_nl();
        while *self.cur() != TT::End {
            out.push(self.stmt());
            self.skip_nl();
        }
        out
    }

    fn block(&mut self) -> Vec<Stmt> {
        self.expect(&TT::Indent);
        let mut out = vec![];
        while *self.cur() != TT::Dedent && *self.cur() != TT::End {
            out.push(self.stmt());
        }
        if *self.cur() == TT::Dedent { self.pos += 1; }
        out
    }

    fn stmt(&mut self) -> Stmt {
        self.skip_nl();
        match self.cur().clone() {
            TT::If    => self.if_stmt(),
            TT::While => {
                self.pos += 1;
                let cond = self.expr(); self.expect(&TT::Colon);
                if *self.cur() == TT::Newline { self.pos += 1; }
                Stmt::While(cond, self.block())
            }
            TT::Def => {
                self.pos += 1;
                let name = if let TT::Ident(n) = self.consume() { n } else { panic!() };
                self.expect(&TT::LParen);
                let mut params = vec![];
                while *self.cur() != TT::RParen {
                    if let TT::Ident(n) = self.consume() { params.push(n); }
                    if *self.cur() == TT::Comma { self.pos += 1; }
                }
                self.expect(&TT::RParen); self.expect(&TT::Colon);
                if *self.cur() == TT::Newline { self.pos += 1; }
                Stmt::Def(name, params, self.block())
            }
            TT::Return => {
                self.pos += 1;
                let e = self.expr();
                if *self.cur() == TT::Newline { self.pos += 1; }
                Stmt::Return(e)
            }
            TT::Print => {
                self.pos += 1; self.expect(&TT::LParen);
                let e = self.expr(); self.expect(&TT::RParen);
                if *self.cur() == TT::Newline { self.pos += 1; }
                Stmt::Print(e)
            }
            TT::Ident(name) if *self.peek() == TT::Assign => {
                self.pos += 2;
                let e = self.expr();
                if *self.cur() == TT::Newline { self.pos += 1; }
                Stmt::Assign(name, e)
            }
            _ => {
                let e = self.expr();
                if *self.cur() == TT::Newline { self.pos += 1; }
                Stmt::Expr(e)
            }
        }
    }

    fn if_stmt(&mut self) -> Stmt {
        self.pos += 1;
        let mut branches = vec![];
        let cond = self.expr(); self.expect(&TT::Colon);
        if *self.cur() == TT::Newline { self.pos += 1; }
        branches.push((cond, self.block()));
        while *self.cur() == TT::Elif {
            self.pos += 1;
            let c = self.expr(); self.expect(&TT::Colon);
            if *self.cur() == TT::Newline { self.pos += 1; }
            branches.push((c, self.block()));
        }
        let else_block = if *self.cur() == TT::Else {
            self.pos += 1; self.expect(&TT::Colon);
            if *self.cur() == TT::Newline { self.pos += 1; }
            self.block()
        } else { vec![] };
        Stmt::If(branches, else_block)
    }

    fn expr(&mut self)    -> Expr { self.or() }
    fn or(&mut self)      -> Expr { self.bin_left(&[TT::Or],      Self::and) }
    fn and(&mut self)     -> Expr { self.bin_left(&[TT::And],     Self::not_expr) }
    fn not_expr(&mut self) -> Expr {
        if *self.cur() == TT::Not { self.pos += 1; return Expr::Unary("not".into(), Box::new(self.not_expr())); }
        self.cmp()
    }
    fn cmp(&mut self) -> Expr {
        let l = self.add();
        let op = match self.cur() {
            TT::Eq => "==", TT::Neq => "!=", TT::Lt => "<",
            TT::Le => "<=", TT::Gt => ">",   TT::Ge => ">=",
            _ => return l,
        }.to_string();
        self.pos += 1;
        Expr::Bin(op, Box::new(l), Box::new(self.add()))
    }
    fn add(&mut self) -> Expr {
        let mut l = self.mul();
        while matches!(self.cur(), TT::Plus | TT::Minus) {
            let op = if *self.cur() == TT::Plus { "+" } else { "-" }.to_string();
            self.pos += 1; l = Expr::Bin(op, Box::new(l), Box::new(self.mul()));
        }
        l
    }
    fn mul(&mut self) -> Expr {
        let mut l = self.unary();
        while matches!(self.cur(), TT::Star | TT::Slash | TT::Percent) {
            let op = match self.cur() { TT::Star => "*", TT::Slash => "/", _ => "%" }.to_string();
            self.pos += 1; l = Expr::Bin(op, Box::new(l), Box::new(self.unary()));
        }
        l
    }
    fn unary(&mut self) -> Expr {
        if *self.cur() == TT::Minus { self.pos += 1; return Expr::Unary("-".into(), Box::new(self.unary())); }
        self.primary()
    }
    fn primary(&mut self) -> Expr {
        match self.cur().clone() {
            TT::Num(n) => { self.pos += 1; Expr::Num(n) }
            TT::Ident(name) => {
                self.pos += 1;
                if *self.cur() == TT::LParen {
                    self.pos += 1;
                    let mut args = vec![];
                    while *self.cur() != TT::RParen {
                        args.push(self.expr());
                        if *self.cur() == TT::Comma { self.pos += 1; }
                    }
                    self.pos += 1;
                    Expr::Call(name, args)
                } else {
                    Expr::Var(name)
                }
            }
            TT::LParen => { self.pos += 1; let e = self.expr(); self.expect(&TT::RParen); e }
            _ => Expr::Num(0),
        }
    }

    fn bin_left(&mut self, ops: &[TT], next: fn(&mut Self) -> Expr) -> Expr {
        let mut l = next(self);
        while ops.iter().any(|o| std::mem::discriminant(o) == std::mem::discriminant(self.cur())) {
            let op = match self.cur() { TT::Or => "or", TT::And => "and", _ => "" }.to_string();
            self.pos += 1; l = Expr::Bin(op, Box::new(l), Box::new(next(self)));
        }
        l
    }
}

// ── Interpreter ───────────────────────────────────────────────────────────────

struct ReturnSignal(i64);

pub struct Interpreter;

impl Interpreter {
    pub fn new() -> Self { Self }

    pub fn run(&mut self, source: &str) -> String {
        let tokens = lex(source);
        let mut parser = Parser { tok: tokens, pos: 0 };
        let program = parser.parse();

        let mut funcs: HashMap<String, (Vec<String>, Vec<Stmt>)> = HashMap::new();
        for s in &program {
            if let Stmt::Def(name, params, body) = s {
                funcs.insert(name.clone(), (params.clone(), body.clone()));
            }
        }

        let mut output = String::new();
        let mut globals: HashMap<String, i64> = HashMap::new();
        let _ = exec_block(&program, &mut globals, &funcs, &mut output);
        output
    }
}

fn exec_block(
    block: &[Stmt],
    env: &mut HashMap<String, i64>,
    funcs: &HashMap<String, (Vec<String>, Vec<Stmt>)>,
    out: &mut String,
) -> Option<ReturnSignal> {
    for s in block {
        if let Some(r) = exec_stmt(s, env, funcs, out) { return Some(r); }
    }
    None
}

fn exec_stmt(
    s: &Stmt,
    env: &mut HashMap<String, i64>,
    funcs: &HashMap<String, (Vec<String>, Vec<Stmt>)>,
    out: &mut String,
) -> Option<ReturnSignal> {
    match s {
        Stmt::Assign(name, expr) => { let v = eval(expr, env, funcs, out); env.insert(name.clone(), v); }
        Stmt::Print(expr)  => { let v = eval(expr, env, funcs, out); out.push_str(&v.to_string()); out.push('\n'); }
        Stmt::Return(expr) => return Some(ReturnSignal(eval(expr, env, funcs, out))),
        Stmt::If(branches, else_block) => {
            for (cond, body) in branches {
                if eval(cond, env, funcs, out) != 0 {
                    let mut local = env.clone();
                    if let Some(r) = exec_block(body, &mut local, funcs, out) {
                        env.extend(local); return Some(r);
                    }
                    env.extend(local); return None;
                }
            }
            let mut local = env.clone();
            if let Some(r) = exec_block(else_block, &mut local, funcs, out) {
                env.extend(local); return Some(r);
            }
            env.extend(local);
        }
        Stmt::While(cond, body) => {
            while eval(cond, env, funcs, out) != 0 {
                let mut local = env.clone();
                if let Some(r) = exec_block(body, &mut local, funcs, out) {
                    env.extend(local); return Some(r);
                }
                env.extend(local);
            }
        }
        Stmt::Def(name, params, body) => {
            let _ = (name, params, body); // already registered
        }
        Stmt::Expr(expr) => { eval(expr, env, funcs, out); }
    }
    None
}

fn eval(
    e: &Expr,
    env: &HashMap<String, i64>,
    funcs: &HashMap<String, (Vec<String>, Vec<Stmt>)>,
    out: &mut String,
) -> i64 {
    match e {
        Expr::Num(n) => *n,
        Expr::Var(name) => *env.get(name).unwrap_or(&0),
        Expr::Unary(op, e) => {
            let v = eval(e, env, funcs, out);
            match op.as_str() { "-" => -v, "not" => if v == 0 { 1 } else { 0 }, _ => v }
        }
        Expr::Bin(op, l, r) => {
            match op.as_str() {
                "and" => { let lv = eval(l, env, funcs, out); if lv == 0 { 0 } else { eval(r, env, funcs, out) } }
                "or"  => { let lv = eval(l, env, funcs, out); if lv != 0 { lv } else { eval(r, env, funcs, out) } }
                _ => {
                    let lv = eval(l, env, funcs, out);
                    let rv = eval(r, env, funcs, out);
                    match op.as_str() {
                        "+"  => lv + rv,
                        "-"  => lv - rv,
                        "*"  => lv * rv,
                        "/"  => lv / rv,
                        "%"  => lv % rv,
                        "==" => (lv == rv) as i64,
                        "!=" => (lv != rv) as i64,
                        "<"  => (lv <  rv) as i64,
                        "<=" => (lv <= rv) as i64,
                        ">"  => (lv >  rv) as i64,
                        ">=" => (lv >= rv) as i64,
                        _    => 0,
                    }
                }
            }
        }
        Expr::Call(name, args) => {
            if let Some((params, body)) = funcs.get(name) {
                let params = params.clone();
                let body = body.clone();
                let mut local: HashMap<String, i64> = params.iter()
                    .zip(args.iter())
                    .map(|(p, a)| (p.clone(), eval(a, env, funcs, out)))
                    .collect();
                if let Some(ReturnSignal(v)) = exec_block(&body, &mut local, funcs, out) {
                    return v;
                }
            }
            0
        }
    }
}
